data/src/main/java/org/apache/iceberg/data/GenericReader.java

```diff
@@ -0,0 +1,292 @@
+  }
+
+  public CloseableIterable<Record> open(CombinedScanTask task) {
+    return new CombinedTaskIterable(task);
+  }
+
+  public CloseableIterable<Record> open(FileScanTask task) {
+    List<DeleteFile> posDeletes = Lists.newArrayList();
+    List<DeleteFile> eqDeletes = Lists.newArrayList();
+    for (DeleteFile delete : task.deletes()) {
+      switch (delete.content()) {
+        case POSITION_DELETES:
+          posDeletes.add(delete);
+          break;
+        case EQUALITY_DELETES:
+          eqDeletes.add(delete);
+          break;
+        default:
+          throw new UnsupportedOperationException("Unknown delete file content: " + delete.content());
+      }
+    }
+
+    Schema fileProjection = fileProjection(!posDeletes.isEmpty());
+
+    CloseableIterable<Record> records = openFile(task, fileProjection);
+    records = applyPosDeletes(records, fileProjection, task.file().path(), posDeletes, task.file());
+    records = applyEqDeletes(records, fileProjection, eqDeletes, task.file());
+    records = applyResidual(records, fileProjection, task.residual());
+
+    return records;
+  }
+
+  private Schema fileProjection(boolean hasPosDeletes) {
+    if (hasPosDeletes) {
+      List<Types.NestedField> columns = Lists.newArrayList(projection.columns());
+      columns.add(MetadataColumns.ROW_POSITION);
+      return new Schema(columns);
+    }
+
+    return projection;
+  }
+
+  private CloseableIterable<Record> applyResidual(CloseableIterable<Record> records, Schema recordSchema,
+                                                  Expression residual) {
+    if (residual != null && residual != Expressions.alwaysTrue()) {
+      InternalRecordWrapper wrapper = new InternalRecordWrapper(recordSchema.asStruct());
+      Evaluator filter = new Evaluator(recordSchema.asStruct(), residual, caseSensitive);
+      return CloseableIterable.filter(records, record -> filter.eval(wrapper.wrap(record)));
+    }
+
+    return records;
+  }
+
+  private CloseableIterable<Record> applyEqDeletes(CloseableIterable<Record> records, Schema recordSchema,
+                                                   List<DeleteFile> eqDeletes, DataFile dataFile) {
+    if (eqDeletes.isEmpty()) {
+      return records;
+    }
+
+    Multimap<Set<Integer>, DeleteFile> filesByDeleteIds = Multimaps.newMultimap(Maps.newHashMap(), Lists::newArrayList);
+    for (DeleteFile delete : eqDeletes) {
+      filesByDeleteIds.put(Sets.newHashSet(delete.equalityFieldIds()), delete);
+    }
+
+    CloseableIterable<Record> filteredRecords = records;
+    for (Map.Entry<Set<Integer>, Collection<DeleteFile>> entry : filesByDeleteIds.asMap().entrySet()) {
+      Set<Integer> ids = entry.getKey();
+      Iterable<DeleteFile> deletes = entry.getValue();
+
+      Schema deleteSchema = TypeUtil.select(recordSchema, ids);
+      int[] orderedIds = deleteSchema.columns().stream().mapToInt(Types.NestedField::fieldId).toArray();
```

Comment:
Does this mean equality deletes can only currently support deletes on top level fields?

Reply:
I intended to support nested fields, but I think there is still a problem here.

For example, schema `1: id bigint, 2: location struct<3: city, 4: postcode>` and delete `postcode='HG1 2AD'`. The delete id set is `{4}`, but the delete schema from the previous line is `2: location struct<4: postcode>`. The projection created from `orderedIds` should use ID 2 because that's the top-level field that needs to be pulled from a full record.

The problem is that the nested struct also needs a projection, which I missed. I think we probably need to build the projection using a visitor and correctly wrap nested structs.